### PR TITLE
Fix stopping of modules started by kubernetes autodiscover

### PR DESCRIPTION
**Pull request by jsoriano**
Kubernetes autodiscover only emits events for containers with
an ID in pods with an IP, but when a pod is being terminated,
their containers can lack of ID and the pod itself can lack of IP.
This leads to modules that are never stopped because the
delete event that should stop them lacks of the needed
information.

This change makes two things to avoid this problem:
* Don't require the pod to have an IP on stop events.
* Use IDs for containers that don't depend on its state.

### CODE REVIEW

File: libbeat/autodiscover/providers/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -113,8 +113,17 @@ func AutodiscoverBuilder(bus bus.Bus, uuid uuid.UUID, c *common.Config) (autodis
...
 		UpdateFunc: func(obj kubernetes.Resource) {
 			logp.Debug("kubernetes", "Watcher Pod update: %+v", obj)
-			p.emit(obj.(*kubernetes.Pod), "stop")
-			p.emit(obj.(*kubernetes.Pod), "start")
+			pod, _ := obj.(*kubernetes.Pod)
+
+			// Avoid restarting a module when update events for a
+			// terminated pod are received
+			if podTerminated(pod) {
+				logp.Debug("kubernetes", "Pod terminated, stopping instead of restarting on update")
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/kubernetes.go]: Not sure, but I think that in deployments where it happens it will happen with frequency.

I am also not sure if the overlap between delete and update events is caused by our implementation or if kubernetes is really sending an update after a delete. This will need more investigation. This code is more defensive that a real fix, I think that no update should arrive or be processed after a delete, not sure why this is happening.

I'd leave this message as debug, the only reason I added it was to know if this path was being executed in case we have to debug it in the future.

File: libbeat/autodiscover/providers/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -125,6 +134,22 @@ func AutodiscoverBuilder(bus bus.Bus, uuid uuid.UUID, c *common.Config) (autodis
...
+	if pod == nil || pod.Metadata == nil {
+		return false
+	}
+	// DeletionTimestamp is protected and can only be set by kubernetes itself.
+	// While this premise is true, we can safely assume that a pod is being
+	// terminated if a value is set.
+	timestamp := pod.Metadata.DeletionTimestamp
+	if timestamp == nil {
+		return false
+	}
```
Comment (jsoriano) [inline libbeat/autodiscover/providers/kubernetes/kubernetes.go]: Added an explanatory comment about the credibility of this check.

**Comment by jsoriano** [comment]
TLDR; My original assumption was wrong, modules were not being re-started because of a race condition between last stop and delete, but they weren't being stopped because on delete we may not have enough information to stop the modules.

On my first implementation of this change I messed up with the `DeletionTimestamp`, I was using the `GetNanos()` method but I have discovered that it always returns 0, so the modules were being deleted as soon as the pod is marked for termination. This fixed the original issue, no modules were running after the pod is stopped, but it was stopping the pods too soon, what can be a problem in some beats.
When I discovered that, I tried to fix it by using Seconds, confirming that all the timestamps were properly calculated. Then I saw the original problem happening again. And then I discovered what was happening.
We only generate events for pods that have an IP, and for containers that have an ID, because we need them, but when a pod is stopping, containers can be dead and without an ID, so now I schedule modules stop as soon as we know that the pod is stopping, at this moment the containers are still alive.

**Comment by jsoriano** [comment]
Writting my previous comment made me think that we may need an identifier for containers that don't depend on its state, so we can start and stop modules normally.

**Comment by jsoriano** [comment]
> Writting my previous comment made me think that we may need an identifier for containers that don't depend on its state, so we can start and stop modules normally.

Yes, much simpler, it works by using another ID, description updated.

**Comment by jsoriano** [comment]
jenkins, test this again please

File: libbeat/autodiscover/providers/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -145,28 +146,23 @@ func (p *Provider) emitEvents(pod *kubernetes.Pod, flag string, containers []*ku
 	host := pod.Status.GetPodIP()
 
 	// Do not emit events without host (container is still being configured)
```
Comment (ruflin) [inline libbeat/autodiscover/providers/kubernetes/kubernetes.go]:
```suggestion
	// Do not emit events without host for containers which are not stopping (container is still being configured)
```
Extending the comment here as I had to read this line multiple times. The part that wondered me here is if should be `flag == start` or something like that to dismiss the start state but not all others. Not sure what the available flags are.

**Comment by ruflin** [comment]
@jsoriano Thanks for the investigation and leaving the thought process trail in the PR, really helpful.

**Comment by ruflin** [comment]
@jsoriano I think failing tests are related. Could be that some tests need update.

File: libbeat/autodiscover/providers/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -145,28 +146,23 @@ func (p *Provider) emitEvents(pod *kubernetes.Pod, flag string, containers []*ku
...
 	for _, c := range containers {
-		cid := containerIDs[c.GetName()]
+		// This must be an id that doesn't depend on the state of the container
+		// so it works also on `stop` if containers have been already stopped.
+		cid := fmt.Sprintf("%s.%s", pod.Metadata.GetUid(), c.GetName())
 
-		// If there is a container ID that is empty then ignore it. It either means that the container is still starting
-		// up or the container is shutting down.
-		if cid == "" {
-			continue
```
Comment (exekias) [inline libbeat/autodiscover/providers/kubernetes/kubernetes.go]: what about the case where the container is still starting?

File: libbeat/autodiscover/providers/kubernetes/kubernetes.go (hunk 3 above)
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/kubernetes.go]: Comments extended.

Regarding flags, we only have `start` and `stop` at the moment, but I only wanted to change the behaviour of stop, and this is why explicitly check for that.

File: libbeat/autodiscover/providers/kubernetes/kubernetes.go (hunk 4 above)
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/kubernetes.go]: I was thinking that just checking for the ID doesn't ensure that the container is ready, so it wouldn't matter so much. But the truth is that there is no need to change this here. I will keep the check for non-stop events.

Thanks for taking the time to review this!

**Comment by jsoriano** [comment]
> @jsoriano I think failing tests are related. Could be that some tests need update.

Except for the adjustement in an event id, failures were legit, I was overwriting also the value for `container.id`, what was wrong. Great to have these tests :slightly_smiling_face:

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
jenkins, test this again

**Comment by jsoriano** [comment]
Oh, this was not the build I wanted to relaunch :man_facepalming:

**Comment by jsoriano** [comment]
jenkins, test this again

**Comment by jsoriano** [comment]
I am merging this to start with the backports, I'll do follow-ups with further fixes if needed.
